### PR TITLE
Feature/flow graph

### PR DESCRIPTION
**Pull request by damianib**
Actually not a flow graph. Don't mind flow_graph.py in the algorithms module.

Compute a restricted set of accessible states by allowing only four types of actions for each player group : 
- Do nothing
- Go to battle against a group of humans or opponents
- Go to battle against two groups of humans/opponents, splitting the player group in two
- Merge with another player group

Cut the number of accessible states by allowing at most one player group per target.
The battle/split behavior can be controlled with two coefficients.

Also fix a few annoying bugs with min-max.

Happy review (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧

### CODE REVIEW

**Comment by damianib** [comment]
refactor done

File: states/calculate_next_restricted.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    if (next_x1, next_y1) == (x2, y2):
+        return [Move(x1, y1, nbMoved1, x2, y2)]
+    else:
+        return [Move(x1, y1, nbMoved1, next_x1, next_y1), Move(x2, y2, nbMoved2, next_x2, next_y2)]
+
+
+# Get minimum number of players to move to attack each target according to coeffBattle
+# coeffBattle = 1 -> no random battle, lower means more risks
+def get_min_nb_moved(dictPlayer: dict, dictOpponent: dict, dictHumans: dict, coeffBattle: float) -> Dict[Tuple[int, int], int]:
+    minMoved = {pos: max(1, floor(1.5 * coeffBattle * nbPawns)) for pos, nbPawns in dictOpponent.items()}
```
Comment (Hackatosh) [inline states/calculate_next_restricted.py]: Shouldn't this be a ceil ?

File: states/calculate_next_restricted.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    minMoved.update({pos: max(1, floor(coeffBattle * nbPawns)) for pos, nbPawns in dictHumans.items()})
```
Comment (Hackatosh) [inline states/calculate_next_restricted.py]: Ceil again no ?

File: states/calculate_next_restricted.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    return minMoved
+
+
+# Get for each player group all the targets it could attack
+def get_all_target_assignments(dictPlayer: dict, dictOpponent: dict, dictHumans: dict, minMoved: dict) -> Dict[Tuple[int, int], List[Tuple[int, int]]]:
```
Comment (Hackatosh) [inline states/calculate_next_restricted.py]: Remove useless arguments

**Comment by Hackatosh** [comment]
get_all_accessible_states_restricted is not the same type as get_all_accessible_states !

File: states/calculate_next_restricted.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                                                          newAssignment)
+                move = get_next_move_to_target(source, target, dictPlayer[source])
+                for recursiveMove in recursiveMoves:
+                    recursiveMove.append(move)
+                    yield recursiveMove
+
+        # choose two targets (split)
+        if nbSplitMax > 0:
+            for (target1, target2) in possibleDoubleTargets:
+                if not (target1 in isAssigned) and not (target2 in isAssigned):
```
Comment (apollisa) [inline states/calculate_next_restricted.py]: See comment of line 89.

File: states/calculate_next_restricted.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    dictHumans = gameState.get_dict_player(HUMANS)
+
+    minMoved = get_min_nb_moved(dictOpponent, dictHumans, coeffBattle)
+    targetAssignments = get_all_target_assignments(dictPlayer, minMoved)
+    doubleTargetAssignments = get_all_double_targets_assignments(dictPlayer, minMoved)
+    # maximum number of splits, coeffSplit = 1 -> no more player groups than targets, higher means more groups
+    nbSplitMax = max(0, ceil(coeffSplit * (len(dictOpponent) + len(dictHumans) - len(dictPlayer))))
+    nbSplitMax = min(len(dictPlayer), nbSplitMax)
+    if not targetAssignments and not doubleTargetAssignments: # checking if coeffBattle wasn't too high
+        minMoved = get_min_nb_moved(dictOpponent, dictHumans, 0)
```
[the remark this answers is not in the document]
Comment (damianib) [reply states/calculate_next_restricted.py]: Yes, but it only happens when the only moves left are suicidal

File: states/calculate_next_restricted.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+        # choose no target
+        recursiveMoves = get_all_assignment_moves(dictPlayer, dictOpponent, dictHumans, minMoved, nbSplitMax,
+                                                  targetAssignments, doubleTargetAssignments, isAssigned)
+        for recursiveMove in recursiveMoves:
+            yield recursiveMove
+
+        # choose one target
+        for target in possibleTargets:
+            if not (target in isAssigned):
```
[the remark this answers is not in the document]
Comment (damianib) [reply states/calculate_next_restricted.py]: Indeed, one target has at most one group in charge of it. 
I was thinking that in order to take down a bigger group, multiple player groups could merge before targeting the big group, hence the merge action.
Your solution will probably work, but complexity-wise if we allow for example two player groups for each target, the number of possible assignments would go roughly from n! to (2n)! (with n the number of targets)

File: states/calculate_next_restricted.py (hunk 4 above)
Comment (damianib) [reply states/calculate_next_restricted.py]: See comment of line 89 :D

File: states/calculate_next_restricted.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (apollisa) [reply states/calculate_next_restricted.py]: Hum okay well in that case, I approve! 😂

**Pull request merged**